ENH: interpolate: replicate `splrep` and `splprep` in Python

ev-br opened the pull request:
#### What does this implement/fix?
<!--Please explain your changes.-->

Implement a clone of `splrep` (== UnivariateSpline)  and `splprep` in python, without FITPACK.

Three objects are added:

- `scipy.interpolate.make_splrep` : a clone of `splrep`
- `scipy.interpolate.make_splprep` : a clone of `splprep`

These two functions have signatures consistent with their FITPACK counterparts, and return a BSpline object, which is consistent with `make_interp_spline`.

- `scipy.interpolate.generate_knots`: this is a generator which constructs knot vectors of increasing length for spline fitting. The core logic replicates that of FITPACK and the code closely follows the Fortran source.

The use of a generator was suggested in https://github.com/scipy/scipy/issues/2579 as alternative to FITPACK's black box routine (see the issue for detailed critique of the existing API).

The `generate_knots` API is a bit lower level: it generates knot vectors only, not smoothing splines. This way, we decouple the knot selection from spline construction, and power users can mix and match various strategies for constructing the spline itself given knots.

One other argument for exposing a generator is that it steps around task={-1, 0, 1} arguments completely (https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.splrep.html).

The generator has two stopping conditions:

- via the smoothing factor s, or
- via the maximum number of knots, nest (the name follows that of FITPACK, maybe worth renaming if user-facing).

The latter is available in the Fortran FITPACK, but effectively disabled in the Fortran-to-python glue layer: https://github.com/scipy/scipy/blob/main/scipy/interpolate/src/fitpack.pyf#L277 or https://github.com/scipy/scipy/blob/main/scipy/interpolate/src/fitpack.pyf#L310.


#### Additional information
<!--Any additional information you think is important.-->

Implementation closely follows that of FITPACK, mainly https://github.com/scipy/scipy/blob/maintenance/1.11.x/scipy/interpolate/fitpack/fpcurf.f and its dependencies.

tylerjereddy commented on scipy/interpolate/_fitpack_repro.py:
```diff
@@ -0,0 +1,823 @@
+            fpart_max = fparts[i]
+
+    if idx_max == -101:
+        raise ValueError("Internal error, please report it to SciPy developers.")
```
Looks like it may be "ok" that this line is hard to hit (not covered by tests at the moment).

ev-br replied:
Yup, this should never be hit (and if it is, we'd need to fix the cause and flip it back to "never be hit").

tylerjereddy commented on scipy/interpolate/_fitpack_repro.py:
```diff
@@ -0,0 +1,823 @@
+    k = operator.index(k)
+
+    if s < 0:
+        raise ValueError(f"`s` must be non-negative. Got {s = }")
```
In case it matters, these 6 errors are not hit by tests yet.

tylerjereddy commented on scipy/interpolate/_fitpack_repro.py:
```diff
@@ -0,0 +1,823 @@
+    # XXX: needed? vs list(generate_knots(...))[-1]
+    for t in generate_knots(x, y, k=k, s=s, w=w, nest=nest, xb=xb, xe=xe):
+        pass
+    return t
```
So far it isn't used by any test, as the comments seem to imply.

tylerjereddy commented on scipy/interpolate/_fitpack_repro.py:
```diff
@@ -0,0 +1,823 @@
+        # [f(p) should be monotonically decreasing]
+        if f1 <= f2 or f2 <= f3:
+            ier, converged = 2, False
+            break
```
fyi, this block hasn't been hit by any of the tests so far

tylerjereddy reviewed (commented):
The test coverage looks pretty good, I added a small number of comments about the few missing lines.

In case I actually end up having to review line-by-line as non-expert, can you clarify if the intent is to replicate the performance of the Fortran implementation as well, or to first just replicate the functionality and then delegate to C/Cython layer as needed to replicate performance later? I see some Cython stuff here already, but not sure if this is ready for i.e., `asv`-style comparison?

j-bowhay commented on scipy/interpolate/_bsplines.py:
```diff
@@ -1534,46 +1542,236 @@ def make_lsq_spline(x, y, t, k=3, w=None, axis=0, check_finite=True):
+        self.offset = offset
+        self.nc = nc
+
+        assert a.ndim == 2
+        assert offset.ndim == 1
+        assert a.shape[0] == offset.shape[0]
```
Are these left over from debugging? Not sure we should have these in the final version as they will be turned off with `-O`

ev-br commented:
Thanks Tyler, Jake! Indeed, the plan is to move parts of this from python to C. I'm doing it piecemeal over at gh-19753, with an intent to sync this one when ready. How much to move to C is still in a bit of a flux: on one hand it's better to keep in python everything which does not strictly need to be in C; on the other hand it gets a bit messy with a mix of C, python and Cython in between. I'll work on this a bit more I guess, to see how this pans out.  

Speaking of reviews --- I'd be much obliged if you could review this one. Please ping me on anything you'd want a non-expert review on!
What would you actually prefer to review --- a three-PR stack or a single one or some other way?

ev-br commented:
I think this is ready from my side. 

Several cleanups are probably needed for the cpp side. At least 
- how to best document cpp routines; is the current way OK, or do we want to go the full doxygen way (hopefully not),
- the current version probably inlines things too aggressively,

That said, I'm going to leave this PR as is, and first take a look at parametric version, `splPrep` instead. 
FITPACK uses two separate routines (https://github.com/scipy/scipy/blob/maintenance/1.11.x/scipy/interpolate/fitpack/fpcurf.f and https://github.com/scipy/scipy/blob/maintenance/1.11.x/scipy/interpolate/fitpack/fppara.f), and I'm wondering how much of the infra could be shared and what really needs to be separate.

So unless there's a review preference, the plan is to defer cleanups to a follow-up, together with the parametric stuff.
Meanwhile this PR can be considered as is, and reviews would be much appreciated.

ev-br commented:
> I'm going to leave this PR as is, and first take a look at parametric version, splPrep instead.

Huh. Implementing splprep turned out to be a great of a dogfooding exercise. The actual change is a dozen of LOC (+ docstrings, tests) _after_ a bit of cleanup of python side internals. So I took an opportunity to do this cleanup and push it here.

ev-br committed:
ENH: interpolate: move bspline/fitpack helpers to a cpp library, add wrappers

ev-br committed:
ENH: interpolate: add a QR solver to make_lsq_spline

ev-br committed:
ENH: interpolate: add make_{splrep,splprep}, gen_knots

FITPACK clones without FITPACK

ev-br commented on scipy/interpolate/_fitpack_repro.py:
```diff
@@ -0,0 +1,983 @@
+        AA[:nc, :nz] = R[:nc, :]
+        # AA[nc:, :] = b.a / p  # done in __call__(self, p)
+        self.AA  = AA
+        self.offset = np.r_[np.arange(nc, dtype=int), b_offset]
```
```suggestion
        self.offset = np.r_[np.arange(nc, dtype=np.intp), b_offset]
```

ev-br committed:
MAINT: try appeasing MyPy and the linter

ev-br commented on scipy/interpolate/src/__fitpack.h:
```diff
@@ -1,64 +1,178 @@
+template<typename T, bool boundscheck=true>
+struct Array1D
+{
+    T* data;
+    ssize_t nelem;
```
The use of this wrapper is to wrap a data pointer from a numpy array. My understanding was that `std::vector` and `ndarray` allocators are incompatible, and std::vector may decide to reallocate the data for its internal reasons. Is this outdated, is it safe to wrap ndarray data pointer in a vector in recent C++ versions?

Kai-Striega replied:
> My understanding was that std::vector and ndarray allocators are incompatible

I'm not 100% sure. I've only ever use Python OR C++, never mixed the two.

> std::vector may decide to reallocate the data for its internal reasons

This is still the case. However, I can't see any examples where this would be triggered (that doesn't mean they don't exist)

ev-br commented on scipy/interpolate/src/__fitpack.cc:
```diff
@@ -0,0 +1,425 @@
+/*
+ *  Find an interval such that t[interval] <= xval < t[interval+1].
+ */
+ssize_t
+_find_interval(const double* tptr, ssize_t len_t,
+               int k,
+               double xval,
+               ssize_t prev_l,
+               int extrapolate)
+{
+    ConstRealArray1D t = ConstRealArray1D(tptr, len_t);
+
+    ssize_t n = t.nelem - k - 1;
+    double tb = t(k);
+    double te = t(n);
+
+    if (xval != xval) {
+        // nan
+        return -1;
+    }
+
+    if (((xval < tb) || (xval > te)) && !extrapolate) {
+        return -1;
+    }
+    ssize_t l = (k < prev_l) && (prev_l < n) ? prev_l : k;
+
+    // xval is in support, search for interval s.t. t[interval] <= xval < t[l+1]
+    while ((xval < t(l)) && (l != k)) {
+        l -= 1;
+    }
+
+    l += 1;
+    while ((xval >= t(l)) && (l != n)) {
+        l += 1;
+    }
+
+    return l-1;
+}
```
IIRC (been a while) there are subtleties on how a binary search handles runs of equal values. Here it would be multiple knots, which are possible at the beginning, at the end and in the middle of the array. I guess it'll need to be carefully tested.

Kai-Striega replied:
That sounds reasonable. If it helps, it might also be worth looking at ``std::upper_bound``

ev-br commented on scipy/interpolate/src/__fitpack.h:
```diff
@@ -1,64 +1,178 @@
+{
+    T* data;
+    ssize_t nelem;
+    T& operator()(const ssize_t i) {
```
While it'd be nice to avoid `operator()` indeed, I thought it'd be nice to be consistent for `Array1D` and `Array2D`? 
OTOH, this whole homemade stuff should probably be replaced with what https://github.com/scipy/scipy/pull/20320 implements. I'll ask over there.

Kai-Striega replied:
Let's focus on the ``std::mdspan`` implementation from here. Do you know if that works well with NumPy arrays?

ev-br commented:
Am tentatively marking this with a 1.14.0 milestone. It's definitely not critical for the release, so can be bumped if it gets on the way. 
The `mdspan` discussion is somewhat orthogonal, we can transition to mdspan as a follow-up (or not transition). Ditto for other `std::` cleanups (std::lower_bound) suggested by Kai.

ev-br committed:
MAINT: interpolate: toggle boundschecking once

ev-br commented on scipy/interpolate/src/__fitpack.h:
```diff
@@ -1,64 +1,178 @@
+typedef Array2D<double, false> RealArray2D;
+typedef Array1D<double, false> RealArray1D;
+typedef Array1D<const double, false> ConstRealArray1D;
+typedef Array2D<const double, false> ConstRealArray2D;
```
Thanks, updated.

Kai-Striega commented:
@ev-br let me know if you would like me to take another look at this. I'll have time this weekend

ev-br commented:
Would be great @Kai-Striega ! 

Also, https://github.com/ev-br/scipy/pull/13 adapts to using mdspan. I checked that the the mods work with both kokkos mdspan and Irwin's version from scipy/special.
The link is a PR to the branch in my scipy fork, so that the diffs with and without mdspan are clearly separate. WDYT?

tylerjereddy commented:
> It's definitely not critical for the release, so can be bumped if it gets in the way.

I'm going to bump it. It would be great if I could be the hero and review everything decisively with a milestone, but not gonna happen on time! Apologies.

ev-br committed:
DOC: interpolate: fix smoke-docs + versionadded

ev-br commented:
Oh the joys of rebasing large PRs after six months. 

A large part of CI failures seem to be due to 

```
E   ImportError: /scipy/build-install/lib/python3.10/site-packages/scipy/interpolate/_bspl.cpython-310-i386-linux-gnu.so: undefined symbol: dlartg_
```

The source of this is me trying to call a LAPAK function from C++. Previously, this was enough:

```
#include "../_build_utils/src/fortran_defs.h"
#define DLARTG F_FUNC(dlartg, DLARTG)
```

but not anymore, apparently.  So I suppose https://github.com/scipy/scipy/issues/20002 stops being theoretical. 

Locally, I can replicate the failure using `python dev.py build --with-scipy-openblas`:

```
In [4]: "dlartg" in dir(_flapack)
Out[4]: True

In [5]: "dlartg_" in dir(_flapack)
Out[5]: False
```

on ubuntu with gcc/gfortran where the name mangling was lowercase with trailing underscore,  `dlartg_`.

Whether this is scipy-openblas or dependent on python/C++ toolchain version, I've no idea.

rgommers commented:
`F_FUNC` isn't right, that just deals with Fortran compiler mangling and upper-case. It does not deal with symbol prefixes/suffixes. The jobs that are failing mostly use `scipy-openblas`, and the symbol name there is `scipy_dlartg_`. I'm out of time for today after gh-21130, but I'll try to work on this `dlartg` example in the coming days.

ev-br committed:
BLD: interpolate: adapt to LAPACK linking from C++

ev-br commented:
Okay, https://github.com/scipy/scipy/pull/19970/commits/63cdbe6038295f064558938c8b79ace1a9bf0336 seems to work locally for `python dev.py build --with-scipy-openblas`. Let's see if full CI concurs.

ev-br committed:
MAINT: interpolate: work around MSVC quirks: 1) ssize_t; 2) std::to_string

ilayn commented:
@ev-br Is this all we need from FITPACK or there will be more to be translated from it?

I can't make heads or tails with this one in terms of the code so apologies for my ignorance.

ilayn commented:
> There will be more @ilhan :-).

![image](https://github.com/user-attachments/assets/0c48969d-9100-4973-be99-18195bb2c557)

> You mean the python translation and its C work routines?

Ah no, I mean the subject matter and the unfortunate Fortran code, FITPACK in general. It looked a bit too cryptic.

ev-br commented:
Well the subject matter is not _that_ bad after all. It's just an LSQ problem in a specific basis + a greedy like heuristics to select a relevant subspace (iow, at iteration m, add a data point with the maximum residual from iteration m-1) + banded linalg.

2D stuff has several tricks on top, which are still to be deciphered, in a sequel not this PR. Most likely, it's something akin to what people do to minimize the bandwidth/fill-ins in sparse linalg. 

Could I tempt you to take a look at the linalg part of what's here now?

ilayn commented:
> 2D stuff has several tricks on top, which are still to be deciphered, in a sequel not this PR. Most likely, it's something akin to what people do to minimize the bandwidth/fill-ins in sparse linalg.

OK surprisingly, that is something I know about. 

> Could I tempt you to take a look at the linalg part of what's here now?

Only if you buy me flowers first 😃  I'll take a look.

ev-br commented:
Can't promise flowers :-). A beverage of your choice yes certainly, next time we're in the same city :-). Meanwhile, need a hand with LAPACK stuff in the lbfgsb PR?

ilayn commented on scipy/interpolate/src/__fitpack.cc:
```diff
@@ -0,0 +1,426 @@
+ * The output `c` array has the shape `(nc, ydim2)`.
+ */
+void
+fpback( /* inputs*/
```
Maybe I read a bit too quickly but isn't [`dtpsv`](https://www.netlib.org/lapack/explore-html/d7/d3b/group__tpsv_gaa7aa06a5635bdd236463a56a889d3aa4.html#gaa7aa06a5635bdd236463a56a889d3aa4) does this since you are using LAPACK?

ev-br replied:
Quite possibly, thanks! My "packed" format is not exactly the LAPACK packed format, so it'll require a bit of work to get right.

ev-br replied:
Actually, no. The "packed" format here only keeps no more than nz=k+2 elements per row, so that the R array here has max size (m*nz) not m(m+1) /2.  (with m >> k). So I can in principle copy, call LAPACK and copy back but then why paying both copying and big-O.

ilayn replied:
Ah OK, so when you mention packed storage, it's not LAPACK's packed storage. Got it.

ilayn commented on scipy/interpolate/src/__fitpack.cc:
```diff
@@ -0,0 +1,426 @@
+
+ */
+void
+qr_reduce(double *aptr, const ssize_t m, const ssize_t nz, // a(m, nz), packed
```
Typically Householder is used for this via [`dlarfg`](https://www.netlib.org/lapack/explore-html/d8/d0d/group__larfg_gadc154fac2a92ae4c7405169a9d1f5ae9.html#gadc154fac2a92ae4c7405169a9d1f5ae9) instead of Givens but it is what it is from the Fortran code. It is OK I guess.

ev-br replied:
Yeah, I left a lengthy comment on how this can be changed into a blocked Householder QR, https://github.com/scipy/scipy/pull/19970/files#diff-8da5866d31e220e1e6d3853b9c7f262a1a034ac88f51f0fc47dcad511397b8a0R201
For the time being, I'd prefer to keep the row-by-row Givens process to follow what FITPACK was doing (and to avoid a ton of bookkeeping, too :-)).

ilayn replied:
It's actually simpler with householder as it holds the reflectors in the same array but probably going rogue in the middle of the original code is not a good move anyways.

ilayn commented:
> Meanwhile, need a hand with LAPACK stuff in the lbfgsb PR?

I won't say no to any help with Fortran 😅 Just give me some more time to make it compilable. I decided to make it a PyModule to bypass Cython.

ev-br commented:
OK, I believe I addressed all review comments so far. 
Refactoring the C++ side can be done in a follow-up. Am going to hit the green button next Monday, Sept 9th, unless somebody wants more time to take a look.

ev-br committed:
TST: interpolate: adapt tests to xp_assert_* infra

ev-br commented:
Okay, let's land this and improve iteratively. Thanks all for the reviews!
